File: skills/scrape-long-url.py
"""
scrape_long_url.py
==================
Hermes dùng script này để cào URL dài (landing page, sales page, etc.)
Gọi Firecrawl API — không cần MCP, không cần VPS, không bị giới hạn độ dài URL.

SETUP:
  export FIRECRAWL_API_KEY="fc-xxxx"   # lấy free tại firecrawl.dev
  python3 scrape_long_url.py "https://100xtourism.com/lp-7ngay.html"

OUTPUT:
  In ra Markdown sạch của trang — AI đọc được ngay.
"""

import urllib.request
import urllib.parse
import json
import sys
import os


FIRECRAWL_API = "https://api.firecrawl.dev/v1/scrape"


def scrape(url: str, api_key: str) -> str:
    """
    Scrape một URL bất kỳ, trả về Markdown sạch.
    Không giới hạn độ dài URL — Firecrawl xử lý hết.
    """
    payload = json.dumps({
        "url": url,
        "formats": ["markdown"],
        "onlyMainContent": True,     # bỏ header/footer/nav rác
        "waitFor": 2000,             # chờ JS render 2 giây (cho landing page)
        "timeout": 30000,
    }).encode("utf-8")

    req = urllib.request.Request(
        FIRECRAWL_API,
        data=payload,
        headers={
            "Authorization": f"Bearer {api_key}",
            "Content-Type": "application/json",
        },
        method="POST",
    )

    try:
        with urllib.request.urlopen(req, timeout=35) as resp:
            result = json.loads(resp.read().decode("utf-8"))

        if result.get("success"):
            return result["data"]["markdown"]
        else:
            return f"[ERROR] Firecrawl trả về lỗi: {result}"

    except urllib.error.HTTPError as e:
        body = e.read().decode("utf-8")
        return f"[HTTP ERROR {e.code}] {body}"
    except Exception as e:
        return f"[ERROR] {type(e).__name__}: {e}"


def scrape_with_meta(url: str, api_key: str) -> dict:
    """
    Scrape + trả về cả metadata (title, description, ogImage...)
    """
    payload = json.dumps({
        "url": url,
        "formats": ["markdown", "extract"],
        "onlyMainContent": True,
        "waitFor": 2000,
        "timeout": 30000,
        "extract": {
            "schema": {
                "type": "object",
                "properties": {
                    "title":       {"type": "string"},
                    "headline":    {"type": "string"},
                    "pain_points": {"type": "array", "items": {"type": "string"}},
                    "benefits":    {"type": "array", "items": {"type": "string"}},
                    "cta":         {"type": "string"},
                    "price":       {"type": "string"},
                }
            }
        }
    }).encode("utf-8")

    req = urllib.request.Request(
        FIRECRAWL_API,
        data=payload,
        headers={
            "Authorization": f"Bearer {api_key}",
            "Content-Type": "application/json",
        },
        method="POST",
    )

    try:
        with urllib.request.urlopen(req, timeout=35) as resp:
            result = json.loads(resp.read().decode("utf-8"))

        if result.get("success"):
            return {
                "markdown": result["data"].get("markdown", ""),
                "metadata": result["data"].get("metadata", {}),
                "extract":  result["data"].get("extract", {}),
            }
        else:
            return {"error": result}

    except Exception as e:
        return {"error": str(e)}


if __name__ == "__main__":
    # Lấy URL từ arg hoặc dùng URL mặc định để test
    url = sys.argv[1] if len(sys.argv) > 1 else "https://100xtourism.com/lp-7ngay.html"
    api_key = os.environ.get("FIRECRAWL_API_KEY", "")

    if not api_key:
        print("[ERROR] Chưa set FIRECRAWL_API_KEY")
        print("Chạy: export FIRECRAWL_API_KEY='fc-xxxx'")
        sys.exit(1)

    print(f"[INFO] Đang cào: {url[:80]}...")
    print("-" * 60)

    # Mode 1: Chỉ lấy markdown
    # markdown = scrape(url, api_key)
    # print(markdown)

    # Mode 2: Lấy markdown + extract structured data
    result = scrape_with_meta(url, api_key)

    if "error" in result:
        print(f"[ERROR] {result['error']}")
    else:
        print("=== EXTRACT (structured) ===")
        print(json.dumps(result["extract"], ensure_ascii=False, indent=2))
        print()
        print("=== MARKDOWN (full content) ===")
        print(result["markdown"][:5000])  # in 5000 ký tự đầu
        if len(result["markdown"]) > 5000:
            print(f"\n... (còn {len(result['markdown']) - 5000} ký tự nữa)")

File: skills/scrape-long-url.js
/**
 * scrape_long_url.js
 * ==================
 * OpenClaw dùng script này để cào URL dài.
 * Gọi Crawl4AI self-hosted (Antigravity đã deploy Docker).
 * Không tốn credit, không giới hạn.
 *
 * SETUP:
 *   Antigravity chạy: docker run -d -p 11235:11235 unclecode/crawl4ai:latest
 *   node scrape_long_url.js "https://100xtourism.com/lp-7ngay.html"
 *
 * FALLBACK: Nếu chưa có Crawl4AI, dùng Firecrawl API (cần key)
 */

const CRAWL4AI_URL = process.env.CRAWL4AI_URL || "http://localhost:11235";
const FIRECRAWL_KEY = process.env.FIRECRAWL_API_KEY || "";

/**
 * Scrape bằng Crawl4AI self-hosted
 * Ưu tiên dùng cái này — free unlimited
 */
async function scrapeWithCrawl4AI(url) {
  const payload = {
    urls: [url],
    crawler_params: {
      headless: true,
      verbose: false,
      wait_for: "body",
      page_timeout: 30000,
    },
    extra: {
      word_count_threshold: 10,
      only_text: false,
    },
  };

  const resp = await fetch(`${CRAWL4AI_URL}/crawl`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });

  if (!resp.ok) {
    throw new Error(`Crawl4AI HTTP ${resp.status}: ${await resp.text()}`);
  }

  const data = await resp.json();

  if (data.results && data.results[0]) {
    return {
      markdown: data.results[0].markdown?.raw_markdown || "",
      html: data.results[0].html || "",
      metadata: data.results[0].metadata || {},
      source: "crawl4ai",
    };
  }

  throw new Error("Crawl4AI: không có kết quả");
}

/**
 * Fallback: Firecrawl API
 * Dùng khi Crawl4AI chưa setup
 */
async function scrapeWithFirecrawl(url, apiKey) {
  const payload = {
    url,
    formats: ["markdown"],
    onlyMainContent: true,
    waitFor: 2000,
    timeout: 30000,
  };

  const resp = await fetch("https://api.firecrawl.dev/v1/scrape", {
    method: "POST",
    headers: {
      Authorization: `Bearer ${apiKey}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify(payload),
  });

  if (!resp.ok) {
    throw new Error(`Firecrawl HTTP ${resp.status}: ${await resp.text()}`);
  }

  const data = await resp.json();

  if (data.success) {
    return {
      markdown: data.data.markdown || "",
      metadata: data.data.metadata || {},
      source: "firecrawl",
    };
  }

  throw new Error(`Firecrawl error: ${JSON.stringify(data)}`);
}

/**
 * Main function — tự chọn engine
 * Thử Crawl4AI trước, fallback sang Firecrawl
 */
async function scrape(url) {
  console.log(`[INFO] Scraping: ${url.substring(0, 80)}...`);

  // Thử Crawl4AI trước (free)
  try {
    const result = await scrapeWithCrawl4AI(url);
    console.log(`[OK] Crawl4AI — ${result.markdown.length} chars`);
    return result;
  } catch (e) {
    console.warn(`[WARN] Crawl4AI failed: ${e.message}`);
  }

  // Fallback Firecrawl
  if (FIRECRAWL_KEY) {
    try {
      const result = await scrapeWithFirecrawl(url, FIRECRAWL_KEY);
      console.log(`[OK] Firecrawl — ${result.markdown.length} chars`);
      return result;
    } catch (e) {
      console.error(`[ERROR] Firecrawl failed: ${e.message}`);
    }
  }

  throw new Error("Cả 2 engine đều fail. Check Crawl4AI Docker hoặc FIRECRAWL_API_KEY.");
}

// CLI entry point
const url = process.argv[2] || "https://100xtourism.com/lp-7ngay.html";

scrape(url)
  .then((result) => {
    console.log("\n=== METADATA ===");
    console.log(JSON.stringify(result.metadata, null, 2));
    console.log("\n=== MARKDOWN (5000 chars) ===");
    console.log(result.markdown.substring(0, 5000));
    if (result.markdown.length > 5000) {
      console.log(`\n... (còn ${result.markdown.length - 5000} chars)`);
    }
  })
  .catch((err) => {
    console.error("[FATAL]", err.message);
    process.exit(1);
  });

// Export cho OpenClaw import
module.exports = { scrape, scrapeWithCrawl4AI, scrapeWithFirecrawl };
